### PR TITLE
FIX(trainer): ensure final checkpoint is saved when resuming training

### LINKED ISSUE #38939: Resuming training from an interrupted checkpoint fails to save the final checkpoint.
### System Info

- `transformers` version: 4.53.0.dev0
- Platform: Linux-4.18.0-147.mt20200626.413.el8_1.x86_64-x86_64-with-glibc2.17
- Python version: 3.12.3
- Huggingface_hub version: 0.30.1
- Safetensors version: 0.5.3
- Accelerate version: 1.6.0
- Accelerate config:    not found
- DeepSpeed version: 0.16.7
- PyTorch version (accelerator?): 2.6.0+cu124 (NA)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?:  yes

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Normally, a trainer that saves checkpoints based on steps or epochs is guaranteed to save a final checkpoint upon completion of the training. This is ensured by the following sequence of logic:

First, when the training reaches the final step, do_sync_step is set to True based on the second part of this conditional statement.

https://github.com/huggingface/transformers/blob/0725cd6953803b8aacfc85288cbfb83dea30c469/src/transformers/trainer.py#L2508

This flag indicates that a gradient update is about to be performed, and the global_step is updated precisely within that gradient update branch. (see L2623)

https://github.com/huggingface/transformers/blob/0725cd6953803b8aacfc85288cbfb83dea30c469/src/transformers/trainer.py#L2569-L2635

This coherent logic would work without any issues in an error-free training run. However, when training is resumed, the step counter is not restored; it is still initialized from -1.

https://github.com/huggingface/transformers/blob/0725cd6953803b8aacfc85288cbfb83dea30c469/src/transformers/trainer.py#L2492

This prevents a gradient update for the final set of batches in an epoch if the total dataset size is not perfectly divisible by the product of gradient_accumulation_steps and the data-parallel world size.

### PR DESCRIPTION
**Pull request by rangehow**
### Fixes

Closes #38939

### Description

This PR addresses an issue where the `Trainer` fails to save the final checkpoint when resuming from an interrupted training run.

The root cause is that the `step` variable within the training loop is always initialized to -1, even when resuming from a checkpoint. This can prevent the gradient update for the last few batches of an epoch if the dataset size is not perfectly divisible by `gradient_accumulation_steps * world_size`. Consequently, the logic to save the final checkpoint is not triggered.

This PR fixes the issue by correctly initializing the `step` variable based on `steps_trained_in_current_epoch` when `resume_from_checkpoint` is used. This ensures that the training loop continues from the exact state it was interrupted, allowing the final gradient update and checkpoint saving to proceed as expected.

I have added a new test case to `tests/trainer/test_trainer_resume.py` to verify this fix. The test simulates resuming from an interrupted training run and asserts that the final checkpoint is correctly saved.

- trainer: @zach-huggingface, @SunMarc and @qgallouedec

### CODE REVIEW

**Comment by rangehow** [comment]
Although there is still work to be done to ensure that resuming the trainer after an interruption is consistent with a full training run (see https://github.com/huggingface/transformers/issues/39755), this PR only addresses the issue of the final checkpoint failing to save. This is because I haven't yet found a good way to handle the saving of the RNG state (unless we consider abandoning the "skip_first_batch" approach a solution).

File: tests/trainer/test_trainer_resume.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+def create_dummy_dataset():
+    """Creates a dummy dataset for testing."""
+    num_samples = 13
+    input_dim = 10
+    dummy_input_ids = torch.rand(num_samples, input_dim)
+    dummy_attention_mask = torch.ones(num_samples, input_dim)
+    dummy_labels = torch.randint(0, 2, (num_samples,))
+    return DummyDictDataset(dummy_input_ids, dummy_attention_mask, dummy_labels)
+
+def test_resume_with_original_trainer():
```
Comment (SunMarc) [inline tests/trainer/test_trainer_resume.py]: let's put that in test_trainer.py under TrainerIntegrationTest class

File: tests/trainer/test_trainer_resume.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    """Tests the original transformers Trainer."""
+    print("Testing the original transformers Trainer...")
+    
+    # 1. Set up a dummy model
+    model = DummyModel(input_dim=10, num_labels=2)
+    dummy_dataset = create_dummy_dataset()
+
+    # 3. First training (simulate interruption)
+    output_dir_initial = "./test_original_trainer_initial"
```
Comment (SunMarc) [inline tests/trainer/test_trainer_resume.py]: instead of creating new dir, you can create a temporary dir by using self.get_auto_remove_tmp_dir()

**Review by SunMarc — COMMENTED** [review]
Thanks for fixing this ! just a few nits. Make sure to fix the style with `make style`

File: tests/trainer/test_trainer_resume.py (hunk 2 above)
Comment (rangehow) [reply tests/trainer/test_trainer_resume.py]: done

File: tests/trainer/test_trainer_resume.py (hunk 1 above)
Comment (rangehow) [reply tests/trainer/test_trainer_resume.py]: done

**Comment by rangehow** [comment]
The issues mentioned above have been resolved. It's ready for another round of code review. 🤗

File: tests/trainer/test_trainer_resume.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        output_dir_initial = self.get_auto_remove_tmp_dir()
+        training_args_initial = TrainingArguments(
+            output_dir=output_dir_initial,
+            num_train_epochs=1,
+            per_device_train_batch_size=2,
+            gradient_accumulation_steps=3,
+            save_strategy="steps",
+            save_steps=1,  # Save at every step
+            report_to=[],  # Disable wandb/tensorboard and other loggers
+            max_steps=2,  # Stop after step 2 to simulate interruption
```
Comment (qgallouedec) [inline tests/trainer/test_trainer_resume.py]: does it really test the described issue? here the training will finish without any error

Comment (rangehow) [reply tests/trainer/test_trainer_resume.py]: Based on my understanding, the test should currently be evaluating the correct trainer, so it shouldn't trigger any error messages. Even with an incorrect trainer, this test code won't report errors during training (which is why this bug is quite subtle). The error check is implemented by verifying whether checkpoint-3 is saved at the end.

Comment (rangehow) [reply tests/trainer/test_trainer_resume.py]: The error manifests as implicitly skipping the training of the last gradient accumulation batch of data, resulting in the failure to save the final checkpoint, rather than causing a crash during training. Please let me know if any part of my description is unclear 😊

File: tests/trainer/test_trainer_resume.py (hunk 1 above)
Comment (SunMarc) [reply tests/trainer/test_trainer_resume.py]: maybe you forgot to push the commit ?

**Review by SunMarc — APPROVED** [review]
Thanks ! Just make sure to move the tests to the file I mentionned !

File: tests/trainer/test_trainer_resume.py (hunk 1 above)
Comment (rangehow) [reply tests/trainer/test_trainer_resume.py]: The code you're looking at seems to be outdated. I've already pushed a new version of the test code, so you might be commenting on the previous one.
In my new code, the test class now inherits from TestCasePlus, because this parent class already includes the logic for finding and cleaning up resources automatically.
I'm not entirely sure what else is needed. From my perspective, the code should be complete now.

Comment (rangehow) [reply tests/trainer/test_trainer_resume.py]: see commit https://github.com/huggingface/transformers/pull/40347/commits/37ff9888a5faa033e5579a42e611f8ad7cdbaed7

Comment (SunMarc) [reply tests/trainer/test_trainer_resume.py]: I mean to put your test code here https://github.com/huggingface/transformers/blob/263d06fedc17bb28f70dabe2acae562bc617ef9b/tests/trainer/test_trainer.py#L1304 
It will be better to not create a new file when introducing new tests

Comment (rangehow) [reply tests/trainer/test_trainer_resume.py]: oh. I see. Thanks for explanation. :hug

**Comment by SunMarc** [comment]
@bot /style

**Comment by rangehow** [comment]
Just a friendly reminder, this PR should be ready to merge now. 🤗

**Comment by SunMarc** [comment]
Thanks for the reminder ! Merging now ;)
